File: utils/json-and-jsonld-utils/src/main/java/de/sovity/edc/utils/jsonld/vocab/Prop.java
/*
 * Copyright (c) 2023 sovity GmbH
 *
 *  This program and the accompanying materials are made available under the
 *  terms of the Apache License, Version 2.0 which is available at
 *  https://www.apache.org/licenses/LICENSE-2.0
 *
 *  SPDX-License-Identifier: Apache-2.0
 *
 *  Contributors:
 *       sovity GmbH - init
 *
 */

package de.sovity.edc.utils.jsonld.vocab;

import lombok.experimental.UtilityClass;

/**
 * Constants for used JSON-LD Vocabulary.
 * <p>
 * Please note, that due to how JSON-LD / ontologies are defined, all property names of a namespace are just
 * mixed together on the same level. A property, e.g. type, might be used in multiple classes, which is an
 * abstraction leak by design.
 */
@UtilityClass
public class Prop {
    public final String ID = "@id";
    public final String TYPE = "@type";
    public final String VALUE = "@value";
    public final String CONTEXT = "@context";
    public final String LANGUAGE = "@language";
    public final String PROPERTIES = "properties";

    @UtilityClass
    public class Edc {
        public final String CTX = "https://w3id.org/edc/v0.0.1/ns/";
        public final String CTX_ALIAS = "edc";
        public final String TYPE_ASSET = CTX + "Asset";
        public final String TYPE_DATA_ADDRESS = CTX + "DataAddress";
        public final String ID = CTX + "id";
        public final String PARTICIPANT_ID = CTX + "participantId";
        public final String PROPERTIES = CTX + "properties";
        public final String PRIVATE_PROPERTIES = CTX + "privateProperties";
        public final String DATA_ADDRESS = CTX + "dataAddress";
        public final String TYPE = CTX + "type";
        public final String DATA_ADDRESS_TYPE_HTTP_DATA = "HttpData";
        public final String DATA_ADDRESS_TYPE_HTTP_PROXY = "HttpProxy";
        public final String BASE_URL = CTX + "baseUrl";
        public final String BODY = CTX + "body";
        public final String METHOD = CTX + "method";
        public final String CONTENT_TYPE = CTX + "contentType";
        public final String QUERY_PARAMS = CTX + "queryParams";
        public final String AUTH_KEY = CTX + "authKey";
        public final String AUTH_CODE = CTX + "authCode";
        public final String SECRET_NAME = CTX + "secretName";
        public final String PROXY_METHOD = CTX + "proxyMethod";
        public final String PROXY_PATH = CTX + "proxyPath";
        public final String PROXY_QUERY_PARAMS = CTX + "proxyQueryParams";
        public final String PROXY_BODY = CTX + "proxyBody";

        // Transfer Request Related
        public static String TYPE_TRANSFER_REQUEST = CTX + "TransferRequest";
        public final String CONNECTOR_ADDRESS = CTX + "connectorAddress";
        public final String CONTRACT_ID = CTX + "contractId";
        public final String CONNECTOR_ID = CTX + "connectorId";
        public final String ASSET_ID = CTX + "assetId";
        public final String DATA_DESTINATION = CTX + "dataDestination";
        public final String RECEIVER_HTTP_ENDPOINT = CTX + "receiverHttpEndpoint";
        public final String MEDIA_TYPE = CTX + "mediaType";
        public final String PATH = CTX + "path";
    }

    /**
     * DCAT Vocabulary, see https://www.w3.org/TR/vocab-dcat-3
     */
    @UtilityClass
    public class Dcat {
        /**
         * Context as specified in https://www.w3.org/TR/vocab-dcat-3/#normative-namespaces
         */
        // TODO: get a sample from the catalog in v0.7 and validate that this is the correct URI and update `catalogResponse.json`
        public final String CTX = "http://www.w3.org/ns/dcat#";

        public final String DATASET = CTX + "dataset";
        public final String DISTRIBUTION = CTX + "distribution";
        public final String DISTRIBUTION_AS_USED_BY_CORE_EDC = CTX + "distribution";
        public final String VERSION = CTX + "version";
        public final String KEYWORDS = CTX + "keyword";
        public final String LANDING_PAGE = CTX + "landingPage";
        public final String MEDIATYPE = CTX + "mediaType";
        public final String START_DATE = CTX + "startDate";
        public final String END_DATE = CTX + "endDate";
        public final String DOWNLOAD_URL = CTX + "downloadURL";
    }

    /**
     * ODRL Vocabulary, see <a href="https://www.w3.org/TR/vocab-dcat-3/">DCAT 3 Specification</a>
     */
    @UtilityClass
    public class Odrl {
        public final String CTX = "http://www.w3.org/ns/odrl/2/";
        public final String HAS_POLICY = CTX + "hasPolicy";
        public final String ACTION = CTX + "action";
        public final String TYPE = CTX + "type";
        public final String CONSTRAINT = CTX + "constraint";
        public final String AND = CTX + "and";
        public final String PERMISSION = CTX + "permission";
        public final String LEFT_OPERAND = CTX + "leftOperand";
        public final String RIGHT_OPERAND = CTX + "rightOperand";
        public final String USE = "USE";
    }

    /**
     * Dcterms Metadata Terms Vocabulary, see <a href="http://purl.org/dc/terms">DCMI Metadata Terms</a>
     */
    @UtilityClass
    public class Dcterms {
        public final String CTX = "http://purl.org/dc/terms/";
        public final String IDENTIFIER = CTX + "identifier";
        public final String TITLE = CTX + "title";
        public final String DESCRIPTION = CTX + "description";
        public final String LANGUAGE = CTX + "language";
        public final String CREATOR = CTX + "creator";
        public final String PUBLISHER = CTX + "publisher";
        public final String LICENSE = CTX + "license";
        public final String TEMPORAL = CTX + "temporal";
        public final String ACCRUAL_PERIODICITY = CTX + "accrualPeriodicity";
        public final String SPATIAL = CTX + "spatial";
        public final String RIGHTS_HOLDER = CTX + "rightsHolder";
        public final String RIGHTS = CTX + "rights";
        public final String RIGHTS_STATEMENT = CTX + "RightsStatement";
    }

    /**
     * Dcterms Metadata Terms Vocabulary, see <a href="https://semantic.sovity.io/dcat-ext/">DCAT 3 Specification</a>
     */
    @UtilityClass
    public class SovityDcatExt {
        public final String CTX = "https://semantic.sovity.io/dcat-ext#";
        public final String CUSTOM_JSON = CTX + "customJson";
        public final String PRIVATE_CUSTOM_JSON = CTX + "privateCustomJson";
        public final String DATA_SOURCE_AVAILABILITY = CTX + "dataSourceAvailability";
        public final String DATA_SOURCE_AVAILABILITY_ON_REQUEST = "ON_REQUEST";
        public final String CONTACT_EMAIL = CTX + "contactEmail";
        public final String CONTACT_PREFERRED_EMAIL_SUBJECT = CTX + "contactPreferredEmailSubject";

        @UtilityClass
        public class HttpDatasourceHints {
            public final String METHOD = CTX + "httpDatasourceHintsProxyMethod";
            public final String PATH = CTX + "httpDatasourceHintsProxyPath";
            public final String QUERY_PARAMS = CTX + "httpDatasourceHintsProxyQueryParams";
            public final String BODY = CTX + "httpDatasourceHintsProxyBody";
        }
    }

    @UtilityClass
    public class SovityMessageExt {
        public final String CTX = "https://semantic.sovity.io/message/generic/";
        public final String REQUEST = CTX + "request";
        public final String RESPONSE = CTX + "response";
        public final String ERROR_MESSAGE = CTX + "errorMessage";
        public final String HEADER = CTX + "header";
        public final String BODY = CTX + "body";
    }

    /**
     * FOAF Vocabulary
     */
    @UtilityClass
    public class Foaf {
        public final String CTX = "http://xmlns.com/foaf/0.1/";
        public final String ORGANIZATION = CTX + "Organization";
        public final String NAME = CTX + "name";
        public final String HOMEPAGE = CTX + "homepage";
    }

    /**
     * Namespace mobilitydcatap as specified in
     * <a href="https://mobilitydcat-ap.github.io/mobilityDCAT-AP/releases/index.html#namespaces">mobilityDCAT-AP</a>
     */
    @UtilityClass
    public class MobilityDcatAp {
        public final String CTX = "https://w3id.org/mobilitydcat-ap/";
        public final String MOBILITY_THEME = CTX + "mobilityTheme";

        @UtilityClass
        public class DataCategoryProps {
            public final String CTX = "https://w3id.org/mobilitydcat-ap/mobility-theme/";
            public final String DATA_CATEGORY = CTX + "data-content-category";
            public final String DATA_SUBCATEGORY = CTX + "data-content-sub-category";
        }

        public final String TRANSPORT_MODE = CTX + "transportMode";
        public final String GEO_REFERENCE_METHOD = CTX + "georeferencingMethod";
        public final String MOBILITY_DATA_STANDARD = CTX + "mobilityDataStandard";

        // Optional property of mobilitydcatap:mobilityDataStandard
        public final String SCHEMA = CTX + "schema";
    }

    /**
     * Namespace skos as specified in
     * <a href="https://mobilitydcat-ap.github.io/mobilityDCAT-AP/releases/index.html#namespaces">mobilityDCAT-AP</a>
     */
    @UtilityClass
    public class Skos {
        public final String CTX = "http://www.w3.org/2004/02/skos/core#";
        public final String PREF_LABEL = CTX + "prefLabel";
    }

    /**
     * Namespace adms as specified in
     * <a href="https://mobilitydcat-ap.github.io/mobilityDCAT-AP/releases/index.html#namespaces">mobilityDCAT-AP</a>
     */
    @UtilityClass
    public class Adms {
        public final String CTX = "http://www.w3.org/ns/adms#";
        public final String SAMPLE = CTX + "sample";
    }

    /**
     * Namespace rdfs as specified in
     * <a href="https://mobilitydcat-ap.github.io/mobilityDCAT-AP/releases/index.html#namespaces">mobilityDCAT-AP</a>
     */
    @UtilityClass
    public class Rdfs {
        public final String CTX = "http://www.w3.org/2000/01/rdf-schema#";
        public final String LITERAL = CTX + "Literal";
        public final String LABEL = CTX + "label";
    }
}

File: connector/ce/libs/jsonld-lib/src/main/kotlin/de/sovity/edc/utils/jsonld/vocab/Prop.kt
/*
 * Copyright 2025 sovity GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Contributors:
 *     sovity - init and continued development
 */
package de.sovity.edc.utils.jsonld.vocab

import lombok.experimental.UtilityClass

/**
 * Constants for used JSON-LD Vocabulary.
 *
 *
 * Please note, that due to how JSON-LD / ontologies are defined, all property names of a namespace are just
 * mixed together on the same level. A property, e.g. type, might be used in multiple classes, which is an
 * abstraction leak by design.
 */
object Prop {
    const val ID: String = "@id"
    const val TYPE: String = "@type"
    const val VALUE: String = "@value"
    const val CONTEXT: String = "@context"
    const val LANGUAGE: String = "@language"
    const val PROPERTIES: String = "properties"

    object Edc {
        const val DSPACE: String = "https://w3id.org/dspace/v0.8/"
        const val CTX: String = "https://w3id.org/edc/v0.0.1/ns/"
        const val CTX_ALIAS: String = "edc"

        const val TYPE_ASSET: String = CTX + "Asset"
        const val TYPE_DATA_ADDRESS: String = CTX + "DataAddress"

        const val ASSET_ID: String = CTX + "assetId"
        const val AUTH_CODE: String = CTX + "authCode"
        const val AUTH_KEY: String = CTX + "authKey"
        const val AUTHORIZATION: String = CTX + "authorization"
        const val BASE_URL: String = CTX + "baseUrl"
        const val BODY: String = CTX + "body"
        const val CONNECTOR_ADDRESS: String = CTX + "connectorAddress"
        const val CONNECTOR_ID: String = CTX + "connectorId"
        const val CONTENT_TYPE: String = CTX + "contentType"
        const val CONTRACT_ID: String = CTX + "contractId"
        const val COUNTER_PARTY_ADDRESS: String = CTX + "counterPartyAddress"
        const val DATA_ADDRESS: String = CTX + "dataAddress"
        const val DATA_ADDRESS_TYPE_HTTP_DATA: String = "HttpData"
        const val DATA_ADDRESS_TYPE_HTTP_PROXY: String = "HttpProxy"
        const val DATA_DESTINATION: String = CTX + "dataDestination"
        const val ENDPOINT: String = CTX + "endpoint"
        const val ID: String = CTX + "id"
        const val MEDIA_TYPE: String = CTX + "mediaType"
        const val METHOD: String = CTX + "method"
        const val PARTICIPANT_ID: String = DSPACE + "participantId"
        const val PATH: String = CTX + "path"
        const val PRIVATE_PROPERTIES: String = CTX + "privateProperties"
        const val PROPERTIES: String = CTX + "properties"
        const val PROXY_BODY: String = CTX + "proxyBody"
        const val PROXY_METHOD: String = CTX + "proxyMethod"
        const val PROXY_PATH: String = CTX + "proxyPath"
        const val PROXY_QUERY_PARAMS: String = CTX + "proxyQueryParams"
        const val QUERY_PARAMS: String = CTX + "queryParams"
        const val RECEIVER_HTTP_ENDPOINT: String = CTX + "receiverHttpEndpoint"
        const val SECRET_NAME: String = CTX + "secretName"
        const val TRANSFER_REQUEST_TRANSFER_TYPE: String = CTX + "transferType"
        const val TYPE: String = CTX + "type"
        const val TYPE_TRANSFER_REQUEST: String = CTX + "TransferRequest"
        const val URL: String = CTX + "url"
        const val ALLOWED_SOURCE_TYPES: String = CTX + "allowedSourceTypes"
        const val ALLOWED_DEST_TYPES: String = CTX + "allowedDestTypes"
        const val ALLOWED_TRANSFER_TYPES: String = CTX + "allowedTransferTypes"
        const val PUBLIC_API_URL: String = CTX + "publicApiUrl"

        object Version07 {
            /**
             * How elements were referenced before the migration to core EDC 0.7.x
             * Preserved for backward compatibility.
             */
            const val CTX: String = "https://w3id.org/edc/v0.0.1/ns/"

            const val PARTICIPANT_ID: String = CTX + "participantId"
        }
    }

    /**
     * DCAT Vocabulary, see https://www.w3.org/TR/vocab-dcat-3
     */
    @UtilityClass
    object Dcat {
        /**
         * Context as specified in https://www.w3.org/TR/vocab-dcat-3/#normative-namespaces
         */
        // TODO: get a sample from the catalog in v0.7 and validate that this is the correct URI and update `catalogResponse.json`
        const val CTX: String = "http://www.w3.org/ns/dcat#"

        const val DATASET: String = CTX + "dataset"

        /**
         * TODO Explain what happen and link code
         */
        const val DISTRIBUTION_WILL_BE_OVERWRITTEN_BY_CATALOG: String = CTX + "distribution"
        const val VERSION: String = CTX + "version"
        const val KEYWORDS: String = CTX + "keyword"
        const val LANDING_PAGE: String = CTX + "landingPage"
        const val MEDIATYPE: String = CTX + "mediaType"
        const val START_DATE: String = CTX + "startDate"
        const val END_DATE: String = CTX + "endDate"
        const val DOWNLOAD_URL: String = CTX + "downloadURL"
    }

    /**
     * ODRL Vocabulary, see [DCAT 3 Specification](https://www.w3.org/TR/vocab-dcat-3/)
     */
    @UtilityClass
    object Odrl {
        const val CTX: String = "http://www.w3.org/ns/odrl/2/"
        const val HAS_POLICY: String = CTX + "hasPolicy"
        const val TYPE: String = CTX + "type"
        const val CONSTRAINT: String = CTX + "constraint"
        const val AND: String = CTX + "and"
        const val USE: String = "USE"
    }

    /**
     * Dcterms Metadata Terms Vocabulary, see [DCMI Metadata Terms](http://purl.org/dc/terms)
     */
    @UtilityClass
    object Dcterms {
        const val CTX: String = "http://purl.org/dc/terms/"
        const val IDENTIFIER: String = CTX + "identifier"
        const val TITLE: String = CTX + "title"
        const val DESCRIPTION: String = CTX + "description"
        const val LANGUAGE: String = CTX + "language"
        const val CREATOR: String = CTX + "creator"
        const val PUBLISHER: String = CTX + "publisher"
        const val LICENSE: String = CTX + "license"
        const val TEMPORAL: String = CTX + "temporal"
        const val ACCRUAL_PERIODICITY: String = CTX + "accrualPeriodicity"
        const val RIGHTS_HOLDER: String = CTX + "rightsHolder"
        const val RIGHTS: String = CTX + "rights"
        const val RIGHTS_STATEMENT: String = CTX + "RightsStatement"
    }

    /**
     * Dcterms Metadata Terms Vocabulary, see [DCAT 3 Specification](https://semantic.sovity.io/dcat-ext/)
     */
    @UtilityClass
    object SovityDcatExt {
        const val CTX: String = "https://semantic.sovity.io/dcat-ext#"
        const val CUSTOM_JSON: String = CTX + "customJson"
        const val PRIVATE_CUSTOM_JSON: String = CTX + "privateCustomJson"
        const val DATA_SOURCE_AVAILABILITY: String = CTX + "dataSourceAvailability"
        const val DATA_SOURCE_AVAILABILITY_ON_REQUEST: String = "ON_REQUEST"
        const val CONTACT_EMAIL: String = CTX + "contactEmail"
        const val CONTACT_PREFERRED_EMAIL_SUBJECT: String = CTX + "contactPreferredEmailSubject"
        const val DISTRIBUTION: String = CTX + "distribution"

        @UtilityClass
        object HttpDatasourceHints {
            const val METHOD: String = CTX + "httpDatasourceHintsProxyMethod"
            const val PATH: String = CTX + "httpDatasourceHintsProxyPath"
            const val QUERY_PARAMS: String = CTX + "httpDatasourceHintsProxyQueryParams"
            const val BODY: String = CTX + "httpDatasourceHintsProxyBody"
        }
    }

    @UtilityClass
    object SovityMessageExt {
        const val CTX: String = "https://semantic.sovity.io/message/generic/"
        const val REQUEST: String = CTX + "request"
        const val RESPONSE: String = CTX + "response"
        const val ERROR_MESSAGE: String = CTX + "errorMessage"
        const val HEADER: String = CTX + "header"
        const val BODY: String = CTX + "body"
    }

    /**
     * FOAF Vocabulary
     */
    @UtilityClass
    object Foaf {
        const val CTX: String = "http://xmlns.com/foaf/0.1/"
        const val ORGANIZATION: String = CTX + "Organization"
        const val NAME: String = CTX + "name"
        const val HOMEPAGE: String = CTX + "homepage"
    }

    /**
     * Namespace mobilitydcatap as specified in
     * [mobilityDCAT-AP](https://mobilitydcat-ap.github.io/mobilityDCAT-AP/releases/index.html#namespaces)
     */
    @UtilityClass
    object MobilityDcatAp {
        const val CTX: String = "https://w3id.org/mobilitydcat-ap/"
        const val MOBILITY_DATA_STANDARD: String = CTX + "mobilityDataStandard"

        // Optional property of mobilitydcatap:mobilityDataStandard
        const val SCHEMA: String = CTX + "schema"
    }

    /**
     * Namespace adms as specified in
     * [mobilityDCAT-AP](https://mobilitydcat-ap.github.io/mobilityDCAT-AP/releases/index.html#namespaces)
     */
    @UtilityClass
    object Adms {
        const val CTX: String = "http://www.w3.org/ns/adms#"
        const val SAMPLE: String = CTX + "sample"
    }

    /**
     * Namespace rdfs as specified in
     * [mobilityDCAT-AP](https://mobilitydcat-ap.github.io/mobilityDCAT-AP/releases/index.html#namespaces)
     */
    @UtilityClass
    object Rdfs {
        const val CTX: String = "http://www.w3.org/2000/01/rdf-schema#"
        const val LITERAL: String = CTX + "Literal"
        const val LABEL: String = CTX + "label"
    }

    @UtilityClass
    object Sphinx {
        const val CTX: String = "https://semantic.sphin-x.de/dcat-ext/"
        const val DATA_MODEL_NAME: String = CTX + "dataModelName"
    }
}
